File: Anuraj-Pariya/q4.py
import math
import os
import random
import re
import sys
class Solution:
    def ispar(self,x):
        stack=[]
        brac= {'{':'}' , '(':')' , '[':']'}
        for char in x:
            if char in ['{','[','(']:
                stack.append(char)
            else:
                if stack:
                    p=stack.pop()
                    if brac[p] != char:
                        return 'NO'
                else:
                    return 'NO'
    
        return 'NO' if stack else 'YES'
 #{ 
#  Driver Code Starts
_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases) :
        #n = int(input())
        #n,k = map(int,imput().strip().split())
        #a = list(map(int,input().strip().split()))
        s = str(input())
        obj = Solution()
        if obj.ispar(s):
            print("balanced")
        else:
            print("not balanced")
# } Driver Code Ends                   
            

File: Pranjal-Singh/q4.cpp
#include<bits/stdc++.h>
using namespace std;


 // } Driver Code Ends



class Solution
{
    public:
    //Function to check if brackets are balanced or not.
    bool ispar(string x)
    { char ch;
     int l=x.length();
     int max=l;
     int top=-1;
     int s[max];
     for(int i=0;i<l;i++)
     {
         if(x[i]=='('||x[i]=='{'||x[i]=='[')
         {s[++top]=x[i];
         continue;}
         if(top==-1)
         return false;
         switch(x[i])
         {
         case ')':
         ch=s[top];
         --top;
         if( ch== '{'||ch== '[')
         return false;
         break;
         case '}':
          ch=s[top];
         --top;
         if( ch=='['||ch=='(')
         return false;
         break;
         case ']':
          ch=s[top];
         --top;
         if( ch=='('||ch=='{')
         return false;
         break;
     }
     }
     if(top<0)
     return true;
     else 
     return false;
     }
     
    

};

// { Driver Code Starts.

int main()
{
   int t;
   string a;
   cin>>t;
   while(t--)
   {
       cin>>a;
       Solution obj;
       if(obj.ispar(a))
        cout<<"balanced"<<endl;
       else
        cout<<"not balanced"<<endl;
   }
}//driver code ends